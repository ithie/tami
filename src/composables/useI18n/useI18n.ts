import { computed, ref, type Ref } from 'vue'

const FALLBACK_LOCALE = 'en'

export const VALID_LOCALES = <const>[FALLBACK_LOCALE, 'de', 'cs', 'it', 'es', 'fr']

export type TValidLocales = (typeof VALID_LOCALES)[number]

const translations: Record<TValidLocales, Record<string, Record<string, string>>> = {
  de: {
    actions: {
      feed: 'Füttern',
      play: 'Spielen',
      sleep: 'Schlafen',
      nothing: 'Nichts'
    },
    labels: {
      hunger: 'Hunger',
      happiness: 'Glück',
      energy: 'Kraft',
      age: 'Alter'
    }
  },
  en: {
    actions: {
      feed: 'Feed',
      play: 'Play',
      sleep: 'Sleep',
      nothing: 'Nothing'
    },
    labels: {
      hunger: 'Hunger',
      happiness: 'Happiness',
      energy: 'Energy',
      age: 'Age'
    }
  },
  cs: {
    actions: {
      feed: 'Nakrmit',
      play: 'Hrát si',
      sleep: 'Spát',
      nothing: 'Nic'
    },
    labels: {
      hunger: 'Hlad',
      happiness: 'Štěstí',
      energy: 'Energie',
      age: 'Věk'
    }
  },
  it: {
    actions: {
      feed: 'Nutrire',
      play: 'Giocare',
      sleep: 'Dormire',
      nothing: 'Niente'
    },
    labels: {
      hunger: 'Fame',
      happiness: 'Felicità',
      energy: 'Energia',
      age: 'Età'
    }
  },
  es: {
    actions: {
      feed: 'Alimentar',
      play: 'Jugar',
      sleep: 'Dormir',
      nothing: 'Nada'
    },
    labels: {
      hunger: 'Hambre',
      happiness: 'Felicidad',
      energy: 'Energía',
      age: 'Edad'
    }
  },
  fr: {
    actions: {
      feed: 'Nourrir',
      play: 'Jouer',
      sleep: 'Dormir',
      nothing: 'Rien'
    },
    labels: {
      hunger: 'Faim',
      happiness: 'Bonheur',
      energy: 'Énergie',
      age: 'Âge'
    }
  }
}

const currentLocale: Ref<TValidLocales> = ref(FALLBACK_LOCALE)

const isValidLocale = (locale: string) => {
  return VALID_LOCALES.includes(locale as unknown as TValidLocales)
}

const setLocale = (locale: TValidLocales = FALLBACK_LOCALE) => {
  currentLocale.value = isValidLocale(locale) ? locale : FALLBACK_LOCALE
}

const currentMessageSet = computed(() => {
  return translations[currentLocale.value]
})

export default () => {
  return {
    setLocale,
    t: (key: string): string => {
      const keys = key.split('.')

      if (!key || keys.length <= 0) {
        return key
      }

      let value = currentMessageSet.value

      for (const k of keys) {
        // @ts-expect-error Type-cast error is okay
        value = value?.[k]
        if (!value) {
          return key
        }
      }

      return value as unknown as string
    }
  }
}
