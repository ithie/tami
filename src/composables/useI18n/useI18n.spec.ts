import { describe, expect, test } from 'vitest'
import { nextTick } from 'vue'
import { mount } from '@vue/test-utils'
import useI18n, { type TValidLocales } from '@/composables/useI18n/useI18n'

const testSets: Record<string, string>[] = [
  { lang: '', translation: 'Feed' },
  { lang: 'de', translation: 'FÃ¼ttern' }
]
describe('useI18n', () => {
  testSets.forEach(({ lang = 'initial', translation }) => {
    test(`${lang} lang translation`, () => {
      const i18n = useI18n()

      if (lang !== 'initial') {
        i18n.setLocale(lang as TValidLocales)
      }

      expect(i18n.t('actions.feed')).toEqual(translation)
    })
    test(`${lang} lang translation but unknown key`, () => {
      const i18n = useI18n()

      expect(i18n.t('un.known.key')).toEqual('un.known.key')
    })
  })
  describe('switches between languages "en" and "de"', () => {
    test('as function', async () => {
      const { setLocale, t } = useI18n()

      setLocale('en' as TValidLocales)

      await nextTick()

      expect(t('actions.sleep')).toEqual('Sleep')

      setLocale('de' as TValidLocales)

      await nextTick()

      expect(t('actions.sleep')).toEqual('Schlafen')
    })
    test('in a rendered component', async () => {
      const { setLocale } = useI18n()

      setLocale('en' as TValidLocales)

      const wrapper = mount({
        template: "<div>{{t('actions.sleep')}}</div>",
        setup() {
          return {
            t: useI18n().t
          }
        }
      })

      expect(wrapper.element).toMatchSnapshot()

      setLocale('de' as TValidLocales)

      await nextTick()

      expect(wrapper.element).toMatchSnapshot()

      setLocale('en' as TValidLocales)

      await nextTick()

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
