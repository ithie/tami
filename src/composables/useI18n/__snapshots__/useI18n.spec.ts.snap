// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useI18n > switches between languages "en" and "de" > in a rendered component 1`] = `
<div>
  Sleep
</div>
`;

exports[`useI18n > switches between languages "en" and "de" > in a rendered component 2`] = `
<div>
  Schlafen
</div>
`;

exports[`useI18n > switches between languages "en" and "de" > in a rendered component 3`] = `
<div>
  Sleep
</div>
`;
